File: codes/Garnetwzy/81.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int s = 0, e = nums.size()-1;
        while(s <= e) {
            int mid = s + (e - s) / 2;
            if(nums[mid] == target)
                return true;
            if(nums[mid] > nums[e]) {
                if(nums[s] <= target && nums[mid] > target) {
                    e = mid - 1;
                }else {
                    s = mid + 1;
                }
            }else if(nums[mid] < nums[e]) {
                if(nums[mid] < target && nums[e] >= target) {
                    s = mid + 1;
                }else{
                    e = mid -1;
                }
            }else{
                e--;
            }
        }
        return false;
    }
};

File: codes/HellDream/81.java
class Solution {
    public boolean search(int[] nums, int target) {
        if(nums==null||nums.length==0) return false;
        int left=0,right=nums.length-1;
        while(left<=right){
            int mid = left+(right-left)/2;
            if(nums[mid]==target) return true;
            if(nums[left]==nums[mid]){
                left++;
            }
            else if(nums[mid]<nums[left]){
                if(target>nums[mid]&&target<=nums[right])
                    left = mid;
                else right = mid-1;
            }
            else{
                if(target<nums[mid]&&target>=nums[left])
                    right = mid;
                else left = mid+1;
            }
        }
        return false;
    }
}